File: console/shell/internal/strings/strings_test.go
package strings

import (
	"reflect"
	stdstrings "strings"
	"testing"
)

func TestVisibleWidthIgnoresANSI(t *testing.T) {
	if got := VisibleWidth("\x1b[31mPrompts\x1b[39m"); got != 7 {
		t.Fatalf("VisibleWidth() = %d, want 7", got)
	}
}

func TestTruncate(t *testing.T) {
	if got := Truncate("Ollin Prompts", 10); got != "Ollin P..." {
		t.Fatalf("Truncate() = %q", got)
	}
}

func TestTruncateClosesANSIStyles(t *testing.T) {
	result := Truncate("\x1b[31mOllin Prompts\x1b[39m", 10)

	if got := ParseAnsiText(result); got != "Ollin P..." {
		t.Fatalf("ParseAnsiText(truncated) = %q", got)
	}
	if !stdstrings.Contains(result, "\x1b[31m") {
		t.Fatalf("expected red ANSI open sequence in %q", result)
	}
	if !stdstrings.Contains(result, "\x1b[0m...") {
		t.Fatalf("expected reset before marker in %q", result)
	}
	if stdstrings.HasSuffix(result, "\x1b[39m") {
		t.Fatalf("did not expect foreground reset suffix in %q", result)
	}
}

func TestTruncateClipsMarkers(t *testing.T) {
	if got := Truncate("Ollin Prompts", 1, "…"); got != "…" {
		t.Fatalf("Truncate(marker) = %q", got)
	}
	if got := Truncate("Ollin Prompts", 1, "東京"); got != "" {
		t.Fatalf("Truncate(wide marker) = %q", got)
	}
}

func TestWrap(t *testing.T) {
	if got := Wrap("Ollin Prompts", 7); !reflect.DeepEqual(got, []string{"Ollin", "Prompts"}) {
		t.Fatalf("Wrap() = %#v", got)
	}

	if got := Wrap("😀", 1); !reflect.DeepEqual(got, []string{"😀"}) {
		t.Fatalf("Wrap(wide) = %#v", got)
	}
}

func TestParseAnsi(t *testing.T) {
	if got := ParseAnsiText("\x1b[32mDone\x1b[39m"); got != "Done" {
		t.Fatalf("ParseAnsiText() = %q", got)
	}

	if got := ParseAnsiSegments("Hello, World!"); !reflect.DeepEqual(got, []AnsiSegment{{Text: "Hello, World!", Codes: ""}}) {
		t.Fatalf("ParseAnsiSegments(plain) = %#v", got)
	}

	want := []AnsiSegment{{Text: "Hello ", Codes: ""}, {Text: "Bold", Codes: "\x1b[1m"}, {Text: " World", Codes: ""}}
	if got := ParseAnsiSegments("Hello \x1b[1mBold\x1b[0m World"); !reflect.DeepEqual(got, want) {
		t.Fatalf("ParseAnsiSegments(styled) = %#v", got)
	}

	want = []AnsiSegment{{Text: "Colored", Codes: "\x1b[38;2;255;100;50m"}}
	if got := ParseAnsiSegments("\x1b[38;2;255;100;50mColored\x1b[0m"); !reflect.DeepEqual(got, want) {
		t.Fatalf("ParseAnsiSegments(rgb) = %#v", got)
	}
}

func TestWrapPreservesANSI(t *testing.T) {
	result := Wrap("\x1b[31mHello World\x1b[0m", 5)

	if len(result) != 2 {
		t.Fatalf("Wrap() len = %d, want 2: %#v", len(result), result)
	}
	for _, want := range []string{"Hello", "World"} {
		found := false
		for _, line := range result {
			found = found || stdstrings.Contains(line, want)
		}
		if !found {
			t.Fatalf("expected wrapped output to contain %q: %#v", want, result)
		}
	}
	for _, line := range result {
		if !stdstrings.Contains(line, "\x1b[31m") || !stdstrings.HasSuffix(line, "\x1b[0m") {
			t.Fatalf("wrapped line does not preserve ANSI: %q", line)
		}
	}
}

File: pkgs-ts/console/tests/strings.test.ts
import { describe, expect, it } from 'vitest';
import { parseAnsiSegments, parseAnsiText, truncate, visibleWidth, wrap } from '#console/index';

describe('string utilities', () => {
	it('measures visible width without ANSI control sequences', () => {
		expect(visibleWidth('\u001B[31mPrompts\u001B[39m')).toBe(7);
	});

	it('truncates by visible width', () => {
		expect(truncate('Ollin Prompts', 10)).toBe('Ollin P...');
	});

	it('closes ANSI styles when truncating styled text', () => {
		const result = truncate('\u001B[31mOllin Prompts\u001B[39m', 10);

		expect(parseAnsiText(result)).toBe('Ollin P...');
		expect(result).toContain('\u001B[31m');
		expect(result).toContain('\u001B[0m...');
		expect(result.endsWith('\u001B[39m')).toBe(false);
	});

	it('clips truncation markers by visible width', () => {
		expect(truncate('Ollin Prompts', 1, '…')).toBe('…');
		expect(truncate('Ollin Prompts', 1, '東京')).toBe('');
	});

	it('wraps words to the requested width', () => {
		expect(wrap('Ollin Prompts', 7)).toEqual(['Ollin', 'Prompts']);
	});

	it('wraps wide characters without emitting empty lines', () => {
		expect(wrap('😀', 1)).toEqual(['😀']);
	});

	it('strips ANSI text', () => {
		expect(parseAnsiText('\u001B[32mDone\u001B[39m')).toBe('Done');
	});

	it('parses plain text into a single ANSI segment', () => {
		expect(parseAnsiSegments('Hello, World!')).toEqual([{ text: 'Hello, World!', codes: '' }]);
	});

	it('parses styled and unstyled ANSI segments', () => {
		expect(parseAnsiSegments('Hello \u001B[1mBold\u001B[0m World')).toEqual([
			{ text: 'Hello ', codes: '' },
			{ text: 'Bold', codes: '\u001B[1m' },
			{ text: ' World', codes: '' },
		]);
	});

	it('parses 24-bit ANSI color segments', () => {
		expect(parseAnsiSegments('\u001B[38;2;255;100;50mColored\u001B[0m')).toEqual([{ text: 'Colored', codes: '\u001B[38;2;255;100;50m' }]);
	});

	it('preserves ANSI codes across wrapped lines', () => {
		const result = wrap('\u001B[31mHello World\u001B[0m', 5);

		expect(result).toHaveLength(2);
		expect(result[0]).toContain('\u001B[31m');
		expect(result[0]).toContain('Hello');
		expect(result[0]?.endsWith('\u001B[0m')).toBe(true);
		expect(result[1]).toContain('\u001B[31m');
		expect(result[1]).toContain('World');
		expect(result[1]?.endsWith('\u001B[0m')).toBe(true);
	});

	it('wraps multiple ANSI color segments onto separate lines', () => {
		const result = wrap('\u001B[31mRed\u001B[0m \u001B[32mGreen\u001B[0m \u001B[34mBlue\u001B[0m', 5);

		expect(result).toHaveLength(3);
		expect(result[0]).toContain('Red');
		expect(result[1]).toContain('Green');
		expect(result[2]).toContain('Blue');
	});
});
